src/par/FixColumnPartitions.c

```diff
@@ -19,138 +19,162 @@
+    ret = 0;
+  }
 
-    if (nfr != 1) {
-      List_Delete(fregs);
-      continue; /* Not a mesh region at the top of the mesh */
-    }
-    
-    /* Hopefully, found the top face of a column in the mesh */
+  List_Delete(rfaces);
+  return ret;
+}
 
-    curreg = List_Entry(fregs,0);
-    curid = MR_ID(curreg);
-    homepid = part[curid-1];
+  
+int FixColumnPartitions(Mesh_ptr mesh, int *part, MSTK_Comm comm) { 
+  int idx, curid, nxtid, homepid, nxtpid, num_cols, num_cells_in_col;
+  int num_touches, num_not_touched;
+  int interior, done, nmove=0;
+  MRegion_ptr mr, curreg, nxtreg;
+  MFace_ptr topf, botf, topf2;
+  List_ptr fregs;
+  int* touched;
+  FILE* fid;
+  
+  if (MESH_Num_Regions(mesh) == 0) return 1; /* Not implemented for 2D (perhaps not needed either) */
+
+  /* ensure every region is touched once */
+  touched = calloc(MESH_Num_Regions(mesh),sizeof(int));
+  for (idx=0; idx!=MESH_Num_Regions(mesh); ++idx) touched[idx] = 0;
+  printf("Read %-d regions", MESH_Num_Regions(mesh));
 
+  /* loop over all regions, finding regions whose "up" face is a boundary */
+  fid = fopen("col_counts.txt", "w");
+
+  num_cols = 0;
+  num_touches = 0;
+  idx = 0;
+  while ((mr = MESH_Next_Region(mesh,&idx))) {
+    FixColumnPartitions_UpDown(mesh, mr, &topf, &botf);
+
+    interior = 1;
+    fregs = MF_Regions(topf);
+    if (List_Num_Entries(fregs) == 1) interior = 0;
     List_Delete(fregs);
 
+    if (interior) continue;
 
+    /* found a top face: iterate to the bottom */
+    num_cols++;
+    num_cells_in_col = 0;
     done = 0;
-    while (!done) {
-
-      rfaces = MR_Faces(curreg);
-      nrf = List_Num_Entries(rfaces);
-      botf = 0;
-      for (i = 0; i < nrf; i++) {
-        rf = List_Entry(rfaces,i);
-
-        fverts = MF_Vertices(rf,1,0);
-        nfv = List_Num_Entries(fverts);
-        found = 0;
-        for (j = 0; j < nfv; j++) {
-          fv = List_Entry(fverts,j);
-          if (MF_UsesEntity(topf,fv,MVERTEX)) {
-            found = 1;  /* found common vertex betweeen two faces */
-            break;
-          }
-        }          
-        List_Delete(fverts);
-
-        if (!found) {
-          botf = rf;
-          break;
-        }
-      }
-      List_Delete(rfaces);
+    curreg = mr;
+    curid = MR_ID(curreg);
+    homepid = part[curid-1];
 
-      /* Found bottom face of current mesh region - get
-       the next lower mesh region and check if it is on
-      the same partition as the top region */
+    /* continue to loop through the column */
+    while (!done) {
+      if (touched[curid-1]) 
+        MSTK_Report("FixColumnPartitions","Mesh is not columnar, iteration process touches the same cell twice.",MSTK_FATAL);
+      touched[curid-1] = 1;
+      num_cells_in_col++;
+      num_touches++;
+      
+      /* find the bottom face of that region */
+      FixColumnPartitions_UpDown(mesh, curreg, &topf2, &botf);
+      if (topf2 != topf) 
```

Comment:
Maybe use "```expected_topf```" and "```topf```" and a call of ```FixColumnPartitions(mesh, curreg, &topf, &botf)```. Right now, the first reaction when we see ```if (topf2 != topf)``` is Whaaaa???